Linux/sgx/test_app/enclave/tests/evp_smx.c

```diff
@@ -465,6 +465,182 @@ int ecall_sm2(void)
 	return ret;
 }
 
+// Encrypt the text
+static int encrypt_sm2(const char* public_key, const unsigned char* plain_text, size_t plain_len, unsigned char** cipher_text, size_t* cipher_len)
+{
+	int ret = 0;
+	size_t buf_len = 0;
+	BIO* pub_bio = NULL;
+	EVP_PKEY* evp_pkey = NULL;
+	EVP_PKEY_CTX* evp_pkey_ctx = NULL;
+
+	// 1. Get SM2 public key using SM2 EVP_PKEY directly
+	pub_bio = BIO_new_mem_buf(public_key, -1);
+	evp_pkey = PEM_read_bio_PUBKEY(pub_bio, &evp_pkey, NULL, NULL);
+	if (evp_pkey == NULL) {
+		printf("Error: fail to get SM2 public key using SM2 EVP_PKEY\n");
+		ret = -1;
+		goto end;
+	}
+
+	evp_pkey_ctx = EVP_PKEY_CTX_new(evp_pkey, NULL);
+	if (evp_pkey_ctx == NULL) {
+		printf("Error: fail to create a new EVP_PKEY_CTX\n");
+		ret = -2;
+		goto end;
+	}
+	if (EVP_PKEY_CTX_set1_id(evp_pkey_ctx, sm2_user_id, sm2_user_id_len) != 1) {
+		printf("Error: fail to set user_id to the EVP_PKEY_CTX\n");
+		ret = -3;
+		goto end;
+	}
+
+	// 2. Encrypt
+	if (EVP_PKEY_encrypt_init(evp_pkey_ctx) <= 0) {
+		printf("Error: fail to init encrypt\n");
+		ret = -4;
+		goto end;
+	}
+	if (EVP_PKEY_encrypt(evp_pkey_ctx, NULL, &buf_len, plain_text, plain_len) <= 0) {
+		printf("Error: fail to calculate buffer length\n");
+		ret = -5;
+		goto end; 
+	}
+	*cipher_text = malloc(buf_len);
+	if (*cipher_text == NULL) {
+		printf("Error: fail to allocate buffer for encrypted text\n");
+		ret = -6;
+		goto end;
+	}
+	memset(*cipher_text, 0, buf_len);
+	if (EVP_PKEY_encrypt(evp_pkey_ctx, *cipher_text, &buf_len, plain_text, plain_len) <= 0) {
+		printf("Error: fail to encrypt\n");
+		ret = -7;
+		goto end;
+	}
+	*cipher_len = buf_len;
+
+end:
+	// 3. Finalize
+	BIO_free(pub_bio);
+	EVP_PKEY_free(evp_pkey);
+	EVP_PKEY_CTX_free(evp_pkey_ctx);
+
+	return ret;
+}
+
+// Decrypt the text
+static int decrypt_sm2(const char* private_key, const char* cipher_text, size_t cipher_len, unsigned char** plain_text, size_t* plain_len)
+{
+	int ret = 0;
+	size_t buf_len = 0;
+	BIO* pri_bio = NULL;
+	EVP_PKEY* evp_pkey = NULL;
+	EVP_PKEY_CTX* evp_pkey_ctx = NULL;
+
+	// 1. Get SM2 private key using SM2 EVP_PKEY directly
+	pri_bio = BIO_new_mem_buf(private_key, -1);
+	evp_pkey = PEM_read_bio_PrivateKey(pri_bio, &evp_pkey, NULL, NULL);
+	if (evp_pkey == NULL) {
+		printf("Error: fail to get SM2 private key using SM2 EVP_PKEY\n");
+		ret = -1;
+		goto end;
+	}
+	
+	evp_pkey_ctx = EVP_PKEY_CTX_new(evp_pkey, NULL);
+	if (evp_pkey_ctx == NULL) {
+		printf("Error: fail to create a EVP_PKEY_CTX\n");
+		ret = -2;
+		goto end;
+	}
+	if (EVP_PKEY_CTX_set1_id(evp_pkey_ctx, sm2_user_id, sm2_user_id_len) != 1) {
```

Comment:
According to the manual of SM2 (https://www.cnblogs.com/zhushoucheng/p/14773804.html), user id looks only affect signature, so no need to set user id when doing encryption. Please check it.